### DISCUSSION TITLE
[NCR#1 Proposal] Better Craigslist with zkEmail and Aztec

### DISCUSSION THREAD

**Discussion by YashBit** [noir-lang/noir#5759] [NRG#1 zkEmail]
#  **A Better Craigslist with ZK Email**

## **Summary:**

Craigslist, established in 1995 by Craig Newmark, began as a local email distribution list in San Francisco and has since evolved into a globally recognized platform for classified ads. The site provides a range of services, from free ads to paid listings in specific categories such as job postings and real estate in select cities. Craigslist primarily facilitates consumer-to-consumer (C2C) transactions, connecting buyers and sellers directly in a community-driven environment where users generate content and community moderation helps maintain quality.

Despite its success, Craigslist faces several issues. Safety concerns are prevalent due to unsecured meet-up locations and problems like ghosting and identity fraud. Current verification methods, including phone number SMS and unencrypted emails, pose significant privacy risks, exposing user data to potential breaches. The introduction of ZK Email could address these concerns by offering enhanced privacy through zero-knowledge proofs, ensuring user information remains confidential and secure.

Furthermore, Craigslist's platform suffers from a clutter of unfiltered listings, poor organization, and limited search functionality. Listings are primarily sorted by date rather than relevance or category, making it challenging for users to locate specific items or services efficiently. We hope to solve these problems in tandem with introducing enhanced security features introduced by zkEmail.

## **Methodology:** 

Google Slides Deck: https://docs.google.com/presentation/d/1DpBjevrsX0JbfOUfDmSt1lKTNH_XKoSt7tbtwGCr96M/edit?usp=sharing

### **User Flow:** 
<p align="center">
  <img src="https://github.com/user-attachments/assets/fc202ce9-fed4-431d-9e1a-db3499ab01c2" alt="Description" />
</p>

#### Step 1.1: User Registration
- A new user (buyer or seller) registers on the platform.

#### Step 1.2: ZK Email Verifier
- The user undergoes KYC verification using the ZK Email Verifier.
- The verification process leverages Zero-Knowledge Proofs (ZKP) to confirm the user's identity without exposing any sensitive information.

#### Step 1.3: Minting ERC-721 Token
- Upon successful KYC verification, an ERC-721 token representing the user's identity is minted and stored in the user's wallet.
- This token serves as a unique, non-fungible identifier for the user on the platform.

#### Step 2.1: Create a Listing
- The seller creates a listing on the platform, associating it with their ERC-721 token.
- The listing details the product or service without including any personal information about the seller.

#### Step 2.2: Anonymous Posting
- The listing is posted anonymously, with only the ERC-721 token linked to it, ensuring the seller’s privacy.
- The platform does not store or display any personal information about the seller.

#### Step 3.1: Search Listings
- A buyer searches for listings on the platform using categories, keywords, or filters.

#### Step 3.2: View Listing Details
- The buyer views the details of a listing without knowing the seller's identity.
- The listing is connected to the seller’s ERC-721 token, but this information is not exposed to the buyer.

#### Step 4.1: Initiate Communication
- The buyer is interested in a listing and initiates communication through the platform.

#### Step 4.2: Encrypted Messaging 
- The platform uses the ERC-721 tokens to facilitate communication.
- Since the platform does not store user data, communication is handled off-chain using encrypted messages.
- Exact Service will be integrated after the MVP has been developed 

#### 1. **Porting Circuits from Circom to Noir**

The initial step is to port existing ZK circuits from Circom to Noir. This transition involves adapting the logic and functionality of the original Circom circuits into Noir's syntax and framework. The circuits being ported include essential components such as those used for email verification and KYC (Know Your Customer) processes. The primary source of these circuits is available at [Circom's email verifier](https://github.com/zkemail/zk-email-verify/blob/main/packages/circuits/email-verifier.circom).

#### 2. **Constructing the ZK Email Verifier from Scratch**

The ZK Email Verifier is a crucial component of this project. It is designed to verify email addresses without exposing sensitive information, leveraging Zero-Knowledge Proofs (ZKP) to ensure privacy. This verifier will be constructed from scratch, incorporating cryptographic techniques including RSA for encryption and SHA for hashing. RSA and SHA are integral to securing the email verification process and ensuring data integrity. The development involves designing the verifier to handle various email verification scenarios while maintaining compliance with privacy standards.

#### 3. **Regex Check for KYC**

The project also includes the development of a regex check for KYC processes, which will be based on the DKIM (DomainKeys Identified Mail) check used in existing KYC checks. This regex check is essential for validating the authenticity of KYC data by ensuring it conforms to predefined patterns and standards. By integrating regex checks, the system can efficiently process and verify KYC information, enhancing overall security and compliance.

#### 4. **Client-Side Proving with Noir**

Client-side proving with Noir is a critical aspect of this methodology. Noir's client-side proving capabilities enable users to generate and submit proofs without exposing their data to third parties. This process involves implementing Noir's proving functions to create zero-knowledge proofs that validate user information and transactions. 

#### 5. **Generating ERC-721 Tokens**

Another significant component is the generation of ERC-721 tokens. ERC-721 is a standard for non-fungible tokens (NFTs) on the Ethereum blockchain, which will be utilized to create unique digital assets associated with verified email addresses or KYC data. 

#### 6. **Integration with Aztec Network**

The Aztec Network will be integrated into the system to enhance privacy and scalability. Aztec provides a solution for privacy-preserving transactions on Ethereum, which will be utilized to ensure that the email verification and KYC processes remain confidential and secure.

### 7. **WebApp** 

A WebApp will be built using NextJS to integrate all the features and create a production ready B2C product. 

#### **Frontend (Next.js):**
  - UI Pages: Home, Listings, Details, Profile
  - Components: Navbar, Footer, Search Bar
  - Static & Server-Side Rendering

#### **Backend Services:**
  -   API Server: Node.js with Express
  -   Authentication: Auth0/Firebase

#### **Data Storage:** 
  - PostgreSQL/MySQL, MongoDB
  - Caching: Redis

#### **Search & Filtering:**
- Search Engine: Elasticsearch/Algolia
- Server-side Filtering

#### **Phase 1: Planning and Setup (Sep 1 - Sep 7, 2024)**

- **Deliverables:**
  - Project plan and milestones
  - Development environment setup
  - Finalized requirements and specifications

#### **Phase 2: Porting and Development (Sep 8 - Oct 6, 2024)**

- **Deliverables:**
  - Circom circuits ported to Noir
  - ZK Email Verifier developed with RSA/SHA integration
  - Regex check for KYC implemented
  - Client-side proving functions in Noir developed and tested

#### **Phase 3: Integration and Testing (Oct 7 - Nov 3, 2024)**

- **Deliverables:**
  - ERC-721 tokens generated and integrated
  - Integration with Aztec Network completed
  - Comprehensive testing of all components and integrations
  - Quality assurance and resolution of issues

#### **Phase 4: Deployment and Review (Nov 4 - Nov 15, 2024)**

- **Deliverables:**
  - Deployment of the ZK Email Verifier and related components
  - Final documentation and user guides completed
  - Collection of stakeholder feedback
  - Review of project outcomes and recommendations for improvements

## **Team:** 
- Yash Bharti 
  - Will be Working Full Time 
  - Experience Writing Noir Circuits with PSE Grant - [Porting over Circuits from Circom -> Noir](https://github.com/privacy-scaling-explorations/zk-kit.noir/pull/8); [Binary Incremental Merkle Tree (Up for Review)](https://github.com/privacy-scaling-explorations/zk-kit.noir/pull/9/files#diff-6813b2670f0706fce45d96c2513d281f1747bc4245e43bed8384d7399635a340)
  - Smart Contracts with Solidity 
  - ZK Circuit Writing : Halo2 (Protostar, IVC Folding Schemes), Circom
  - Proficient with Rust, C++, Python, Web Dev - NextJS
  - NYU BA CS '22, University of Cambridge - MPhil (CBL Lab) - Deferred 2025
  - AI Research Background - Deep Reinforcement Learning for Robotics, Natural Language Understanding, Cognitive Science
  - Telegram: yb3351, Email: yashbharti924@gmail.com

## **Start Date:**
- 1 September, 2024

**Comment by Savio-Sou** [comment]
Hi @YashBit, the proposal looks good and thanks for submitting it!

A few early questions and thoughts:

### Craiglist's problems and solutions

> identity fraud / exposing user data to potential breaches

Is it right to understand the aim here as leveraging zkEmails to zkKYC users, hence removing the need to store user data in the first place?

> unsecured meet-up locations / unfiltered listings / poor organization / limited search functionality

Could you elaborate on how do zkEmails solve these issues?

### ERC-721 / integration with Aztec Network

Is the aim to deploy KYC NFTs on the Aztec DevNet directly, or to deploy them on Ethereum then bridge onto the Aztec DevNet?

(If it's the former, deploying verifiers is likely unnecessary where deploying Aztec.nr contracts would be more appropriate instead; if it's the latter, all good.)

### Start Date

The selection committee will strive to finalize decisions as soon as possible once submission closes, no later than 5 September 2024.

Minor note that despite the possibility, there is no guarantee that the committee will announce before the date.

That said, early explorations are welcome anytime.

**Reply by YashBit** [reply]
@Savio-Sou Thank you for the comments  

> Is it right to understand the aim here as leveraging zkEmails to zkKYC users, hence removing the need to store user data in the first place?

Yes, the core idea is to lower the decision tax for users to create accounts and post. The current methods need you to share an email + receive a SMS from a phone number. 

zkKYC users can create and login to accounts whose information will not be stored into the company server 

> Could you elaborate on how do zkEmails solve these issues?

I was mostly thinking of first KYC, then creating metrics in the future for a user based on previous successful transactions. Also, a credit score verification. With regards to locations, I was thinking of implementing a "safe-lock" feature where the location of the meet-up is decided by a verified third party to enhance security (I need to think about this more). 

> Is the aim to deploy KYC NFTs on the Aztec DevNet directly, or to deploy them on Ethereum then bridge onto the Aztec DevNet?

My motivation is to use the security features of Aztec DevNet. I want to be able to add an extra layer of comfort to the users mind when they register for this service. 

What would you recommend? My thesis was that such an approach would only enrich the Aztec ecosystem.

The entire point is to leverage Aztec's brand (For eg, market using - Powered by Aztec), to create a sense of perfect security and trust in the service. 



Thank you again. Quite excited about this proposition. Would appreciate your feedback, and would love to clear any doubts that you may have.

**Comment by Savio-Sou** [comment]
Selected proposals for NCR#1 are announced: https://github.com/orgs/noir-lang/discussions/5932

Thank you for your support on the initiative. NCR is just one of the many ways to start your explorations with Noir, check the announcement for more potential ways to collaborate.

Let's bring the world ZK!
